public/app/plugins/datasource/prometheus/querybuilder/components/PromQueryBuilderContainer.tsx

```diff
@@ -17,25 +17,40 @@ export interface Props {
   data?: PanelData;
 }
 
+export interface State {
+  visQuery?: PromVisualQuery;
+  expr: string;
+}
+
 /**
  * This component is here just to contain the translation logic between string query and the visual query builder model.
- * @param props
- * @constructor
  */
 export function PromQueryBuilderContainer(props: Props) {
   const { query, onChange, onRunQuery, datasource, data } = props;
+  const [state, setState] = useState<State>({ expr: query.expr });
 
-  const visQuery = buildVisualQueryFromString(query.expr || '').query;
+  // Only rebuild visual query if expr changes from outside
+  useEffect(() => {
+    if (!state.visQuery || state.expr !== query.expr) {
+      const result = buildVisualQueryFromString(query.expr || '');
+      setState({ visQuery: result.query, expr: query.expr });
+    }
+  }, [query.expr, state.visQuery, state.expr]);
```

Comment:
I don't know, this will work I think, but feel bad about having 3 pieces of state that represent the same thing and need syncing. You mentioned:
> The other way to solve it is inside OperationEditor but it's not clear how it should know what is a valid restParam and when it should be added to the model vs not.

Wouldn't just having more types easier to handle? Like having param.type = "nonempty:string" or something that would tell the input to run onChange after there is something in the input and otherwise just handle it in local state? I don't think I will be able to dig deeper into the code today, so not sure if that would actually be simpler though.

Reply:
Tried change OperationEditor but it's very hard as adding a rest param changes the whole operation from sum to a sum by, and handling all that with transient (local) component state is tricky, as all of a sudden the operation state that comes from parent needs to be ignored and the local uncommitted state needs to be processed.

I have tried but OperationEditor becomes so much more complex and messy requires special duplicated handling for the uncommitted rest parameter, as well as the whole changed operation that is uncommitted

Reply:
Managed to find a way to make it work, https://github.com/grafana/grafana/pull/45419

Reply:
Ok yeah I think this probably makes more sense than the operation approach and will also scale to other parts of the query when needed.